src/dtls13.c

```diff
@@ -0,0 +1,1968 @@
+{
+    Dtls13RecordNumber *rn, *list;
+
+    WOLFSSL_ENTER("Dtls13RtxAddAck");
+
+    rn = Dtls13NewRecordNumber(ssl, recordNumber);
+    if (rn == NULL)
+        return MEMORY_E;
+
+    list = fsm->ackRecords;
+    if (list == NULL) {
+        fsm->ackRecords = rn;
+        return 0;
+    }
+
+    while (list->next != NULL)
+        list = list->next;
+
+    list->next = rn;
+
+    return 0;
+}
+
+static void Dtls13RtxFlushAcks(WOLFSSL *ssl, Dtls13RtxFSM *fsm)
+{
+    Dtls13RecordNumber *list, *rn;
+
+    (void)ssl;
+
+    WOLFSSL_ENTER("Dtls13RtxFlushAcks");
+
+    list = fsm->ackRecords;
+
+    while (list != NULL) {
+        rn = list;
+        list = rn->next;
+        XFREE(rn, ssl->heap, DYNAMIC_TYEP_DTLS_MSG);
+    }
+
+    fsm->ackRecords = NULL;
+}
+
+static int Dtls13DetectDisruption(WOLFSSL *ssl, word32 fragOffset)
+{
+    /* retranmisssion. The other peer may have lost our flight or our ACKs. We
+       not account this as a distruption */
+    if (ssl->keys.dtls_peer_handshake_number <
+        ssl->keys.dtls_expected_peer_handshake_number)
+        return 0;
+
+    /* out of order message */
+    if (ssl->keys.dtls_peer_handshake_number >
+        ssl->keys.dtls_expected_peer_handshake_number)
+        return 1;
+
+    /* first fragment of in-order message */
+    if (fragOffset == 0)
+        return 0;
+
+    /* is not the next fragment in the message (the check is not 100% perfect,
+       in the worst case, we don't detect the disruption and wait for the other
+       peer retransmission) */
+    if (ssl->dtls_rx_msg_list == NULL ||
+        ssl->dtls_rx_msg_list->fragSz != fragOffset)
+        return 1;
+
+    return 0;
+}
+
+static void Dtls13RtxFSMSetState(
+    WOLFSSL *ssl, Dtls13RtxFSM *fsm, enum Dtls13RtxFsmState state)
+{
+    if (fsm->state == state)
+        return;
+
+    fsm->state = state;
+
+    switch(state) {
+    case DTLS13_RTX_FSM_PREPARING:
+    case DTLS13_RTX_FSM_FINISHED:
+        Dtls13RtxFlushBuffered(ssl, fsm);
+        break;
+    case DTLS13_RTX_FSM_SENDING:
+        Dtls13RtxFlushAcks(ssl, fsm);
+        break;
+    case DTLS13_RTX_FSM_WAITING:
+        break;
+    default:
+        return;
+    }
+}
+
+static int Dtls13RtxIsLastFligh(WOLFSSL *ssl)
+{
+    (void)ssl;
+
+#ifndef NO_WOLFSSL_SERVER
+    if (ssl->options.side == WOLFSSL_SERVER_END &&
+        ssl->options.acceptState >= TLS13_ACCEPT_FINISHED_SENT)
+        return 1;
+#endif /* NO_WOLFSSL_SERVER */
+
+#ifndef NO_WOLFSSL_CLIENT
+    if (ssl->options.side == WOLFSSL_CLIENT_END &&
+            ssl->options.connectState >= WAIT_FINISHED_ACK)
+        return 1;
+#endif
+
+    return 0;
+}
+
+static void Dtls13RtxRecordRecvd(WOLFSSL *ssl, enum HandShakeType hs,
+    Dtls13RtxFSM *fsm, word32 recordNumber[2], word32 fragOffset)
+{
+    int ret;
+
+    WOLFSSL_ENTER("Dtls13RtxRecordRecvd");
+
+    ret = Dtls13RtxAddAck(ssl, fsm, recordNumber);
+    if (ret != 0) {
+        WOLFSSL_MSG("can't save ack fragment");
+    }
```

Comment:
Not erroring out here could potentially cause deadlocks where we won't send the require ACKs. The peer will keep sending its last flight and will not receive an ACK for it ever.